### ISSUE TITLE
Scheduler plugin takes Nodes' Pods CIDR in into account

### ISSUE THREAD

**Issue by lianghao208** [kubernetes-sigs/scheduler-plugins#199]
Is that possible to take nodes' Pods CIDR in into account in scheduling phase?
I have met this situation multiple times:
Pods were bond to Nodes without enough Pod IPs to assign.
It leads to a fatal problem that Pods can't run normally since there was no IP to assign.
So is that possible to check if Nodes has enough Pod IPs to assign in scheduler plugin?

**Comment by denkensk** [comment]
> Pods were bond to Nodes without enough Pod IPs to assign.
---

Yes. This is common when the density of pods in a node is relatively high. We set the MaxPod of node as the limit of pods   to address this issue. But I think we can also try to solve it in scheduler.

**Comment by lianghao208** [comment]
> > Pods were bond to Nodes without enough Pod IPs to assign.
> 
> Yes. This is common when the density of pods in a node is relatively high. We set the MaxPod of node as the limit of pods to address this issue. But I think we can also try to solve it in scheduler.

@denkensk Can I solve this issue by implementing a PreFilter plugin to ignore the nodes that have no IP to allocate( Simply determine if the pod already used up all the PodCIDR field from nodes resource).

**Comment by denkensk** [comment]
Yes. I think it makes sense.  You can start with a simple proposal for your idea.

**Comment by wangyanghack** [comment]
There is situation that some cni plugin using ENI network interface to get the performance equivalent to the underlying network. Quantity of pods in a node depending on quantity of ENI and quantity of ip every ENI can be attached. So is that possible to check if Nodes has enough ENIs and free IPs on ENI  to assign in scheduler plugin?

**Comment by lianghao208** [comment]
> So is that possible to check if Nodes has enough ENIs and free IPs on ENI to assign in scheduler plugin?
Yes, I think it needs some places to store each node's allocatable IPs for pods, either in node's `status.allocatable` field or in a CRD, WDYT?

**Comment by wangyanghack** [comment]
> > So is that possible to check if Nodes has enough ENIs and free IPs on ENI to assign in scheduler plugin?
> 
> @wangyanghack Yes, I think it needs some places to store each node's allocatable IPs for pods, either in node's `status.allocatable` field or in a CRD, WDYT?

@lianghao208 Yes, I agree with you. Checking nodes in cluster who use ENIs cannot use .spec.PodCIDR to calculate node's allocatable IPs for pods.
